veps/sig-compute/55-evictionstrategy-extension.md

```diff
@@ -0,0 +1,154 @@
+
+# Design
+This section outlines the theory behind how this design works and the code changes required to implement this theory into practice.
+
+## Design Theory
+### Actual behavior
+#### Scheduler Priority Eviction
+
+When a pod cannot be scheduled due to constraints, the scheduler may trigger preemption.
+If the incoming pod has a high priority, the scheduler will preempt lower-priority pods by setting a condition `Type: v1.DisruptionTarget`, `Reason: v1.PodReasonPreemptionByScheduler`, then delete the pod directly (not using the eviction API).
+Currently, this causes a graceful shutdown of the VM, not a live migration—even if `evictionStrategy: LiveMigrate` is set.
+See: https://github.com/kubernetes/kubernetes/blob/3e5849972e619c58b3a8e8be2ca3ac35c4eb74da/pkg/scheduler/framework/preemption/preemption.go#L160-L202
+
+Also, the scheduler can bypass PDB protection when necessary.
+See: https://github.com/kubernetes/kubernetes/blob/3e5849972e619c58b3a8e8be2ca3ac35c4eb74da/pkg/scheduler/framework/preemption/preemption.go#L678-L679
+
+#### `NoExecute` Taint Eviction
+
+When a node is tainted with `Effect: NoExecute`, the taint manager evicts all non-tolerating pods.
+It sets a condition with `Type: v1.DisruptionTarget`, `Reason: "DeletionByTaintManager"`, then deletes the pod.
+See: https://github.com/kubernetes/kubernetes/blob/3e5849972e619c58b3a8e8be2ca3ac35c4eb74da/pkg/controller/tainteviction/taint_eviction.go#L129-L148
+
+This, again, causes a graceful shutdown instead of a live migration.
+PDBs are ignored by the taint manager.
+
+## Repos
+
+- `kubevirt/kubevirt` – main implementation
+- `kubevirt/user-guide` – documentation updates
+- `kubevirt/hyperconverged-operator` – exposure and enablement of the feature gate
+
+## Design
+
+`virt-controller` already watches `virt-launcher` pods.
+If a pod has a `deletionTimestamp` but the corresponding VMI does not, the controller should:
+
+1. Check for `DisruptionTarget` condition with:
+   - `Status == True`
+   - `Reason == DeletionByTaintManager` or `PodReasonPreemptionByScheduler`
+2. Treat this as a valid eviction request and follow the VMI’s `EvictionStrategy`.
+
+Grace period handling remains consistent with existing eviction logic.
+
+## API Examples
+
+These behaviors are guarded by opt-in feature gates:
+- `PreemptionEvictionLiveMigration`
+- `NoExecuteTaintEvictionLiveMigration`
+
+Once enabled, KubeVirt responds to these eviction causes based on the per-VM or cluster-wide `EvictionStrategy`.
+
+## Alternatives
+
+The ideal long-term solution is for the scheduler and taint manager to use the eviction API.
+However, these are sensitive components so such a change could require a long path. Thus, this proposal can act as a consistent workaround in the meantime.
+In the past, something around ideas in this area got discussed in https://github.com/kubernetes/kubernetes/issues/91492 
```

Comment:
[KEP-3280](https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/3280-guarantee-pdb-when-preemption-happens) is also mentioning:
```
if the priority of the preemptor is greater than or equal to the value of AllowDisruptionByPriorityGreaterThanOrEqual in victim pod, the implementation will remain consistent with the existing behavior, meaning that the scheduler will try to select victims whose PDBs are not violated by preemption, but if no such victims are found, preemption will still happen and lower priority pods will be preempted via the /evictions endpoint despite their PDBs being violated. If the /eviction endpoint returns a response 429 Too Many Requests and the scheduler will fallback to deletion as an alternative.
```